FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1148 @@
+        } else {
+          CmpNode = IREmit->_F80Cmp(StackNode, Value);
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80XTRACTSTACK: {
+        auto Value = MigrateToSlowPath_IfInvalid();
+        Ref St0 = LoadStackValue(Value);
+
+        Ref Exp {};
+        Ref Sig {};
+        if (ReducedPrecisionMode) {
+          std::tie(Exp, Sig) = SplitF64SigExp(St0);
+        } else {
+          Exp = IREmit->_F80XTRACT_EXP(St0);
+          Sig = IREmit->_F80XTRACT_SIG(St0);
+        }
+
+        if (SlowPath) {
+          // Write exp to top, update top for a push and set sig at new top.
+          StoreStackValueAtTop_Slow(Exp, false);
+          UpdateTopForPush_Slow();
+          StoreStackValueAtTop_Slow(Sig);
+        } else {
+          StackData.setTop(StackMemberInfo {Exp});
+          StackData.push(StackMemberInfo {Sig});
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_SYNCSTACKTOSLOW: {
+        // This synchronizes stack values but doesn't necessarily moves us off the FastPath!
+        Ref NewTop = SynchronizeStackValues();
+        IREmit->ReplaceUsesWithAfter(CodeNode, NewTop, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_STACKFORCESLOW: {
+        MigrateToSlowPathIf(true);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_INCSTACKTOP:
+      case OP_DECSTACKTOP: {
+        if (SlowPath) {
+          if (IROp->Op == OP_DECSTACKTOP) {
+            UpdateTopForPop_Slow();
+          } else {
+            UpdateTopForPush_Slow();
+          }
+        } else {
+          StackData.rotate(IROp->Op == OP_DECSTACKTOP);
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80ROUNDSTACK: {
+        auto StackMember = MigrateToSlowPath_IfInvalid();
+        Ref St0 = LoadStackValue(StackMember);
+
+        Ref Value {};
+        if (ReducedPrecisionMode) {
+          Value = IREmit->_Vector_FToI(8, 8, St0, Round_Host);
+        } else {
+          Value = IREmit->_F80Round(St0);
+        }
+        StoreStackValue(Value);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80VBSLSTACK: {
+        const auto* Op = IROp->C<IROp_F80VBSLStack>();
+
+        // Multiplies two elements in the stack by offset.
+        auto* VecCond = CurrentIR.GetNode(Op->VectorMask);
+        auto StackOffset1 = Op->SrcStack1;
+        auto StackOffset2 = Op->SrcStack2;
+        auto ValueOffset1 = MigrateToSlowPath_IfInvalid(StackOffset1);
+        auto ValueOffset2 = MigrateToSlowPath_IfInvalid(StackOffset2);
+        Ref Value1 = LoadStackValue(ValueOffset1, StackOffset1);
+        Ref Value2 = LoadStackValue(ValueOffset2, StackOffset2);
+
+        Ref StackNode = IREmit->_VBSL(16, VecCond, Value1, Value2);
+        StoreStackValue(StackNode, 0, StackOffset1 && StackOffset2);
```

Comment:
I don't understand the && here.

Reply:
I guess it's `StackOffset1 != 0 && StackOffset2 != 0`. The point is, we are storing to the top of the stack. We only really need to mark it as valid if none of the values come from the top of the stack. Because if they did, we would know it's already valid and there's no point in re-marking it.